# Foundations of Computer Science 2023 Project

We first import useful packages

In [1]:
import pandas as pd
import numpy as np
import re

## 1. Extract all trips with busRoute 83

First of all we need to read the dataset

In [2]:
trips=pd.read_csv('metaData.csv')

Let's check what the dataset is like

In [3]:
trips

name  busNumber  \
0     B183_2019-04-30_03-18-56_2019-04-30_08-44-20        183   
1     B183_2019-04-30_13-22-07_2019-04-30_17-54-02        183   
2     B183_2019-05-01_05-58-51_2019-05-01_22-32-30        183   
3     B183_2019-05-03_02-50-21_2019-05-03_05-53-20        183   
4     B183_2019-05-03_15-41-57_2019-05-03_23-06-24        183   
...                                            ...        ...   
1404  B208_2022-12-06_14-43-49_2022-12-06_18-22-52        208   
1405  B208_2022-12-07_05-13-02_2022-12-07_19-19-53        208   
1406  B208_2022-12-08_05-22-20_2022-12-08_18-39-15        208   
1407  B208_2022-12-09_23-55-12_2022-12-10_03-24-28        208   
1408  B208_2022-12-10_23-34-46_2022-12-11_03-29-05        208   

             startTime_iso  startTime_unix           endTime_iso  \
0     2019-04-30T03:18:56Z      1556594336  2019-04-30T08:44:20Z   
1     2019-04-30T13:22:07Z      1556630527  2019-04-30T17:54:02Z   
2     2019-05-01T05:58:51Z      1556690331  2019-05-01T22:32:30Z   
3     2019-05-03T02:50:21Z      1556851821  2019-05-03T05:53:20Z   
4     2019-05-03T15:41:57Z      1556898117  2019-05-03T23:06:24Z   
...                    ...             ...                   ...   
1404  2022-12-06T14:43:49Z      1670337829  2022-12-06T18:22:52Z   
1405  2022-12-07T05:13:02Z      1670389982  2022-12-07T19:19:53Z   
1406  2022-12-08T05:22:20Z      1670476940  2022-12-08T18:39:15Z   
1407  2022-12-09T23:55:12Z      1670630112  2022-12-10T03:24:28Z   
1408  2022-12-10T23:34:46Z      1670715286  2022-12-11T03:29:05Z   

      endTime_unix  drivenDistance busRoute  energyConsumption  \
0       1556613860        77213.87        -       4.785852e+08   
1       1556646842        59029.60       31       4.022585e+08   
2       1556749950       240900.40       33       1.445733e+09   
3       1556862800        42565.48        -       2.819867e+08   
4       1556924784       125277.20       72       6.207258e+08   
...            ...             ...      ...                ...   
1404    1670350972        51798.78       32       4.260419e+08   
1405    1670440793       210041.60       83       1.536697e+09   
1406    1670524755       190372.70       83       1.415700e+09   
1407    1670642668        59548.57       N1       4.519165e+08   
1408    1670729345        72293.83       N2       5.350956e+08   

      itcs_numberOfPassengers_mean  itcs_numberOfPassengers_min  \
0                         5.538860                            0   
1                        33.114580                            4   
2                        19.689140                            0   
3                         1.685185                            0   
4                        23.753570                            1   
...                            ...                          ...   
1404                     39.808990                            0   
1405                     28.785390                            0   
1406                     29.879400                            0   
1407                     20.105260                            0   
1408                     15.993200                            0   

      itcs_numberOfPassengers_max  status_gridIsAvailable_mean  \
0                              20                     0.740640   
1                              74                     0.855234   
2                              55                     0.777860   
3                               8                     0.767122   
4                              67                     0.907342   
...                           ...                          ...   
1404                           83                     0.739349   
1405                          115                     0.434858   
1406                          102                     0.439916   
1407                           74                     0.495739   
1408                           52                     0.459492   

      temperature_ambient_mean  temperature_ambient_min  

Let us check what the different routes were

In [4]:
trips['busRoute']

0        -
1       31
2       33
3        -
4       72
        ..
1404    32
1405    83
1406    83
1407    N1
1408    N2
Name: busRoute, Length: 1409, dtype: object

Let's see if it recognizes 83 

In [5]:
trips['busRoute']==83

0       False
1       False
2       False
3       False
4       False
        ...  
1404    False
1405    False
1406    False
1407    False
1408    False
Name: busRoute, Length: 1409, dtype: bool

Since it doesn't (see lines 1405-1406) and there are routes named with letters, the data might be stored in strings

In [6]:
trips['busRoute']=='83'

0       False
1       False
2       False
3       False
4       False
        ...  
1404    False
1405     True
1406     True
1407    False
1408    False
Name: busRoute, Length: 1409, dtype: bool

Ok, that seems to be the case.

In [7]:
trips[trips['busRoute']=='83']

name  busNumber  \
154   B183_2020-03-03_04-42-38_2020-03-03_19-44-51        183   
155   B183_2020-03-06_04-53-23_2020-03-06_19-44-42        183   
157   B183_2020-03-09_14-16-13_2020-03-09_19-34-17        183   
158   B183_2020-03-10_04-50-03_2020-03-10_19-51-25        183   
159   B183_2020-03-12_04-56-41_2020-03-12_19-44-57        183   
...                                            ...        ...   
1399  B208_2022-11-30_04-47-53_2022-11-30_19-50-22        208   
1400  B208_2022-12-01_05-19-41_2022-12-01_18-20-57        208   
1401  B208_2022-12-02_04-47-48_2022-12-02_19-40-01        208   
1405  B208_2022-12-07_05-13-02_2022-12-07_19-19-53        208   
1406  B208_2022-12-08_05-22-20_2022-12-08_18-39-15        208   

             startTime_iso  startTime_unix           endTime_iso  \
154   2020-03-03T04:42:38Z      1583210558  2020-03-03T19:44:51Z   
155   2020-03-06T04:53:23Z      1583470403  2020-03-06T19:44:42Z   
157   2020-03-09T14:16:13Z      1583763373  2020-03-09T19:34:17Z   
158   2020-03-10T04:50:03Z      1583815803  2020-03-10T19:51:25Z   
159   2020-03-12T04:56:41Z      1583989001  2020-03-12T19:44:57Z   
...                    ...             ...                   ...   
1399  2022-11-30T04:47:53Z      1669783673  2022-11-30T19:50:22Z   
1400  2022-12-01T05:19:41Z      1669871981  2022-12-01T18:20:57Z   
1401  2022-12-02T04:47:48Z      1669956468  2022-12-02T19:40:01Z   
1405  2022-12-07T05:13:02Z      1670389982  2022-12-07T19:19:53Z   
1406  2022-12-08T05:22:20Z      1670476940  2022-12-08T18:39:15Z   

      endTime_unix  drivenDistance busRoute  energyConsumption  \
154     1583264691       225047.90       83       1.544278e+09   
155     1583523882       224512.30       83       1.631816e+09   
157     1583782457        77824.36       83       5.406013e+08   
158     1583869885       225095.80       83       1.692171e+09   
159     1584042297       224181.20       83       1.145860e+09   
...            ...             ...      ...                ...   
1399    1669837822       223165.00       83       1.560888e+09   
1400    1669918857       190196.00       83       1.418847e+09   
1401    1670010001       224473.40       83       1.611150e+09   
1405    1670440793       210041.60       83       1.536697e+09   
1406    1670524755       190372.70       83       1.415700e+09   

      itcs_numberOfPassengers_mean  itcs_numberOfPassengers_min  \
154                       23.47531                            0   
155                       17.41578                            0   
157                       23.18182                            0   
158                       20.96410                            0   
159                       17.21235                            0   
...                            ...                          ...   
1399                      27.89066                            2   
1400                      26.03927                            0   
1401                      24.80384                            2   
1405                      28.78539                            0   
1406                      29.87940                            0   

      itcs_numberOfPassengers_max  status_gridIsAvailable_mean  \
154                           118                     0.472180   
155                            69                     0.451028   
157                            74                     0.460099   
158                            86                     0.475233   
159                            80                     0.340882   
...                           ...                          ...   
1399                          100                     0.456196   
1400                           96                     0.450413   
1401                           91                     0.438693   
1405                          115                     0.434858   
1406                          102                     0.439916   

      temperature_ambient_mean  temperature_ambient_min  

## 2. Extract all trips where busRoute is not a number

Let us define a function to check if the route is not a number. We have to do this because some routes are just missing (see lines 0 or 3 as example) and some others are letters and numbers (such as 1407-1408)

In [8]:
def check_not_number(route):
    m=re.match('^\d+\d$' ,route)
    if m:
        return False
    else: return True

Let's test it on some strings 

In [9]:
l='lol'
a='83'
b='N1'
c='189492749274398273'
d='18273t87217hh198718198'

In [10]:
check_not_number(l)

True

In [11]:
check_not_number(a)

False

In [12]:
#check_not_number(p)

In [13]:
check_not_number(b)

True

In [14]:
check_not_number(c)

False

In [15]:
check_not_number(d)

True

In [16]:
last_one='12N3'

In [17]:
check_not_number(last_one)

True

Since it seems to work we can use it on the actual dataset

In [18]:
trips[trips['busRoute'].apply(check_not_number)==True]

name  busNumber  \
0     B183_2019-04-30_03-18-56_2019-04-30_08-44-20        183   
3     B183_2019-05-03_02-50-21_2019-05-03_05-53-20        183   
9     B183_2019-05-10_03-16-11_2019-05-10_18-51-37        183   
10    B183_2019-05-13_03-10-23_2019-05-13_23-16-13        183   
19    B183_2019-05-24_02-52-47_2019-05-24_22-35-11        183   
...                                            ...        ...   
1373  B208_2022-10-21_22-38-32_2022-10-22_02-42-21        208   
1374  B208_2022-10-22_22-34-45_2022-10-23_02-29-59        208   
1394  B208_2022-11-25_23-35-16_2022-11-26_03-30-39        208   
1407  B208_2022-12-09_23-55-12_2022-12-10_03-24-28        208   
1408  B208_2022-12-10_23-34-46_2022-12-11_03-29-05        208   

             startTime_iso  startTime_unix           endTime_iso  \
0     2019-04-30T03:18:56Z      1556594336  2019-04-30T08:44:20Z   
3     2019-05-03T02:50:21Z      1556851821  2019-05-03T05:53:20Z   
9     2019-05-10T03:16:11Z      1557458171  2019-05-10T18:51:37Z   
10    2019-05-13T03:10:23Z      1557717023  2019-05-13T23:16:13Z   
19    2019-05-24T02:52:47Z      1558666367  2019-05-24T22:35:11Z   
...                    ...             ...                   ...   
1373  2022-10-21T22:38:32Z      1666391912  2022-10-22T02:42:21Z   
1374  2022-10-22T22:34:45Z      1666478085  2022-10-23T02:29:59Z   
1394  2022-11-25T23:35:16Z      1669419316  2022-11-26T03:30:39Z   
1407  2022-12-09T23:55:12Z      1670630112  2022-12-10T03:24:28Z   
1408  2022-12-10T23:34:46Z      1670715286  2022-12-11T03:29:05Z   

      endTime_unix  drivenDistance busRoute  energyConsumption  \
0       1556613860        77213.87        -       4.785852e+08   
3       1556862800        42565.48        -       2.819867e+08   
9       1557514297       210577.00        -       1.303391e+09   
10      1557789373       267033.80        -       1.647432e+09   
19      1558737311       263432.60        -       1.448057e+09   
...            ...             ...      ...                ...   
1373    1666406541        78567.16       N1       4.347766e+08   
1374    1666492199        73427.97       N2       3.997737e+08   
1394    1669433439        72911.26       N2       4.475534e+08   
1407    1670642668        59548.57       N1       4.519165e+08   
1408    1670729345        72293.83       N2       5.350956e+08   

      itcs_numberOfPassengers_mean  itcs_numberOfPassengers_min  \
0                         5.538860                            0   
3                         1.685185                            0   
9                         8.230483                            0   
10                        7.891652                            0   
19                        7.520249                            0   
...                            ...                          ...   
1373                     16.333330                            0   
1374                     17.710530                            0   
1394                     11.216670                            1   
1407                     20.105260                            0   
1408                     15.993200                            0   

      itcs_numberOfPassengers_max  status_gridIsAvailable_mean  \
0                              20                     0.740640   
3                               8                     0.767122   
9                              43                     0.740927   
10                             45                     0.804191   
19                             44                     0.761068   
...                           ...                          ...   
1373                           45                     0.431852   
1374                           57                     0.443358   
1394                           32                     0.465024   
1407                           74                     0.495739   
1408                           52                     0.459492   

      temperature_ambient_mean  temperature_ambient_min  

## 3. For each (busNumber, busRoute) pair, determine the number of trips

Let's check how many different buses number and routes we have:

In [19]:
trips['busNumber'].unique()

array([183, 208], dtype=int64)

In [20]:
trips['busRoute'].unique()

array(['-', '31', '33', '72', '46', '32', '83', 'N4', 'N2', 'N1'],
      dtype=object)

Now we can just group by Number and Route and check how many lines we count

In [21]:
trips.groupby(['busNumber', 'busRoute']).size().reset_index(name='Number of trips').drop(0)

busNumber busRoute  Number of trips
1         183       31               12
2         183       32               12
3         183       33              130
4         183       46              104
5         183       72              114
6         183       83              441
7         183       N1               10
8         183       N2               19
9         183       N4               11
10        208       31                5
11        208       32               14
12        208       33               25
13        208       46               19
14        208       72               44
15        208       83              405
16        208       N1                6
17        208       N2               20
18        208       N4                7

This is in accord with point 2 of this project: the total number of trips for Route 83 is 441 for bus 183 and 405 for bus 208; that is a total of 846 trips, as the lines resulting in "trips[trips['busRoute']=='83']"

## 4. For each trip, compute the ratio between the energy consumption and the average number of passengers 

First of all let's check the data type of energyConsumption: if it is a string we'd have to convert it to a number

In [22]:
type(trips['energyConsumption'][0])

numpy.float64

No work needed here, we can just add a column with the requested information

In [23]:
trips['EnergyRatio']=trips['energyConsumption']/trips['itcs_numberOfPassengers_mean']

In [24]:
trips['EnergyRatio']

0       8.640500e+07
1       1.214747e+07
2       7.342794e+07
3       1.673328e+08
4       2.613190e+07
            ...     
1404    1.070215e+07
1405    5.338462e+07
1406    4.738047e+07
1407    2.247753e+07
1408    3.345769e+07
Name: EnergyRatio, Length: 1409, dtype: float64

## 5. For each station (itcs_stopName), determine the average number of passengers. 

itcs_stopName does not ring a bell but let's check if it is in the columns

In [25]:
import zipfile

In [26]:
zf = zipfile.ZipFile('ZTBUS_compressed.zip','r')

In [27]:
print(zf.namelist())
#questo ritorna una lista di stringhe con i nomi 

['B183_2019-04-30_03-18-56_2019-04-30_08-44-20.csv', 'B183_2019-04-30_13-22-07_2019-04-30_17-54-02.csv', 'B183_2019-05-01_05-58-51_2019-05-01_22-32-30.csv', 'B183_2019-05-03_02-50-21_2019-05-03_05-53-20.csv', 'B183_2019-05-03_15-41-57_2019-05-03_23-06-24.csv', 'B183_2019-05-05_07-41-02_2019-05-05_23-20-07.csv', 'B183_2019-05-06_03-10-43_2019-05-06_19-20-34.csv', 'B183_2019-05-08_02-54-58_2019-05-08_23-00-52.csv', 'B183_2019-05-09_03-20-44_2019-05-09_22-52-12.csv', 'B183_2019-05-10_03-16-11_2019-05-10_18-51-37.csv', 'B183_2019-05-13_03-10-23_2019-05-13_23-16-13.csv', 'B183_2019-05-14_03-33-11_2019-05-14_12-48-45.csv', 'B183_2019-05-14_13-58-18_2019-05-14_21-04-54.csv', 'B183_2019-05-15_03-04-56_2019-05-15_06-58-44.csv', 'B183_2019-05-15_13-31-00_2019-05-15_22-29-33.csv', 'B183_2019-05-18_07-41-27_2019-05-18_23-02-16.csv', 'B183_2019-05-20_03-07-29_2019-05-20_17-27-37.csv', 'B183_2019-05-21_03-41-23_2019-05-21_22-45-31.csv', 'B183_2019-05-22_02-58-32_2019-05-22_23-02-32.csv', 'B183_2019-

In [28]:
nomi_file = trips['name']
nomi_file

0       B183_2019-04-30_03-18-56_2019-04-30_08-44-20
1       B183_2019-04-30_13-22-07_2019-04-30_17-54-02
2       B183_2019-05-01_05-58-51_2019-05-01_22-32-30
3       B183_2019-05-03_02-50-21_2019-05-03_05-53-20
4       B183_2019-05-03_15-41-57_2019-05-03_23-06-24
                            ...                     
1404    B208_2022-12-06_14-43-49_2022-12-06_18-22-52
1405    B208_2022-12-07_05-13-02_2022-12-07_19-19-53
1406    B208_2022-12-08_05-22-20_2022-12-08_18-39-15
1407    B208_2022-12-09_23-55-12_2022-12-10_03-24-28
1408    B208_2022-12-10_23-34-46_2022-12-11_03-29-05
Name: name, Length: 1409, dtype: object

In [29]:
dataframe = []

In [30]:
#with zipfile.ZipFile('ZTBUS_compressed.zip', 'r') as zip_ref:
#    for nome_file in nomi_file:
#        percorso_file_zip = nome_file+'.csv'
#        with zip_ref.open(percorso_file_zip) as file_zip:
#            df = pd.read_csv(file_zip)
#            dataframe.append(df)
#df_completo = pd.concat(dataframe, ignore_index=True)

In [31]:
trips.columns

Index(['name', 'busNumber', 'startTime_iso', 'startTime_unix', 'endTime_iso',
       'endTime_unix', 'drivenDistance', 'busRoute', 'energyConsumption',
       'itcs_numberOfPassengers_mean', 'itcs_numberOfPassengers_min',
       'itcs_numberOfPassengers_max', 'status_gridIsAvailable_mean',
       'temperature_ambient_mean', 'temperature_ambient_min',
       'temperature_ambient_max', 'EnergyRatio'],
      dtype='object')

let's check if by chance it's in the other dataset...

In [32]:
test=pd.read_csv('ZTBus/B183_2019-04-30_03-18-56_2019-04-30_08-44-20.csv')

In [33]:
test.head(480)

time_iso   time_unix  electric_powerDemand  gnss_altitude  \
0    2019-04-30T03:18:56Z  1556594336            -13.845510            NaN   
1    2019-04-30T03:18:57Z  1556594337             -3.849362            NaN   
2    2019-04-30T03:18:58Z  1556594338             -0.672331            NaN   
3    2019-04-30T03:18:59Z  1556594339             -1.087931            NaN   
4    2019-04-30T03:19:00Z  1556594340             -0.811985            NaN   
..                    ...         ...                   ...            ...   
475  2019-04-30T03:26:51Z  1556594811        -136126.400000            NaN   
476  2019-04-30T03:26:52Z  1556594812        -157050.400000            NaN   
477  2019-04-30T03:26:53Z  1556594813         -90376.580000            NaN   
478  2019-04-30T03:26:54Z  1556594814         -52053.430000            NaN   
479  2019-04-30T03:26:55Z  1556594815         -48089.290000            NaN   

     gnss_course  gnss_latitude  gnss_longitude itcs_busRoute  \
0            NaN            NaN             NaN             -   
1            NaN            NaN             NaN             -   
2            NaN            NaN             NaN             -   
3            NaN            NaN             NaN             -   
4            NaN            NaN             NaN             -   
..           ...            ...             ...           ...   
475          NaN            NaN             NaN             -   
476          NaN            NaN             NaN             -   
477          NaN            NaN             NaN             -   
478          NaN            NaN             NaN             -   
479          NaN            NaN             NaN             -   

     itcs_numberOfPassengers itcs_stopName  ...  odometry_wheelSpeed_mr  \
0                        NaN             -  ...                0.000000   
1                        NaN             -  ...                0.000000   
2                        NaN             -  ...                0.000000   
3                        NaN             -  ...                0.000000   
4                        NaN             -  ...                0.000000   
..                       ...           ...  ...                     ...   
475                      NaN             -  ...               10.012080   
476                      NaN             -  ...                8.908013   
477                      NaN             -  ...                8.190138   
478                      NaN             -  ...                7.402411   
479                      NaN             -  ...                6.676773   

     odometry_wheelSpeed_rl  odometry_wheelSpeed_rr  status_doorIsOpen  \
0                  0.000000                0.000000                  1   
1                  0.000000                0.000000                  1   
2                  0.000000                0.000000                  1   
3                  0.000000                0.000000                  1   
4                  0.000000                0.000000                  1   
..                      ...                     ...                ...   
475               10.041520               10.027170                  0   
476                9.001329                9.001329                  0   
477                8.224860                8.207500                  0   
478                7.416729                7.382008                  0   
479                6.700973                6.704480                  0   

     status_gridIsAvailable  status_haltBrakeIsActive  \
0                         1                         0   
1                         1                         0   
2                         1                         0   
3                         1                         0   
4                         1                         0   
..                      ...                       ...   
475                       1                         0   
476                       1                         0

In [34]:
test['itcs_stopName'].unique()

array(['-', 'Zürich, Herdernstrasse', 'Zürich, Hardplatz',
       'Zürich, Güterbahnhof', 'Zürich, Bäckeranlage',
       'Zürich, Militär-/Langstrasse', 'Zürich, Kanonengasse',
       'Zürich, Sihlpost / HB', 'Zürich, Löwenplatz',
       'Zürich, Bahnhofplatz/HB', 'Zürich, Central', 'Zürich, Neumarkt',
       'Zürich, Kunsthaus', 'Zürich, Sprecherstrasse',
       'Zürich, Kreuzplatz', 'Zürich, Signaustrasse',
       'Zürich, Hegibachplatz', 'Zürich, Botanischer Garten',
       'Zürich, Höschgasse', 'Zürich, Fröhlichstrasse',
       'Zürich, Wildbachstrasse', 'Zürich, Bahnhof Tiefenbrunnen',
       'Zürich, Freiestrasse', 'Zürich, Klusplatz',
       'Zürich, Hölderlinsteig', 'Zürich, Klosbach', 'Zürich, Hofstrasse',
       'Zürich, Kirche Fluntern', 'Zürich, Hinterbergstrasse',
       'Zürich, Spyriplatz', 'Zürich, Bethanien', 'Zürich, Toblerplatz'],
      dtype=object)

In [35]:
len(test['itcs_stopName'].unique())

32

In [36]:
nome=trips['name']
nome

0       B183_2019-04-30_03-18-56_2019-04-30_08-44-20
1       B183_2019-04-30_13-22-07_2019-04-30_17-54-02
2       B183_2019-05-01_05-58-51_2019-05-01_22-32-30
3       B183_2019-05-03_02-50-21_2019-05-03_05-53-20
4       B183_2019-05-03_15-41-57_2019-05-03_23-06-24
                            ...                     
1404    B208_2022-12-06_14-43-49_2022-12-06_18-22-52
1405    B208_2022-12-07_05-13-02_2022-12-07_19-19-53
1406    B208_2022-12-08_05-22-20_2022-12-08_18-39-15
1407    B208_2022-12-09_23-55-12_2022-12-10_03-24-28
1408    B208_2022-12-10_23-34-46_2022-12-11_03-29-05
Name: name, Length: 1409, dtype: object

In [37]:
nome=trips['name']

a=nome[0]
a='ZTBUS/'+a+'.csv'
a

'ZTBUS/B183_2019-04-30_03-18-56_2019-04-30_08-44-20.csv'

In [38]:
tr=pd.read_csv(a, usecols=['itcs_stopName'])

In [39]:
tr

itcs_stopName
0                 -
1                 -
2                 -
3                 -
4                 -
...             ...
19520             -
19521             -
19522             -
19523             -
19524             -

[19525 rows x 1 columns]

In [40]:
tr[tr['itcs_stopName']!='-']

itcs_stopName
246          Zürich, Herdernstrasse
322               Zürich, Hardplatz
406            Zürich, Güterbahnhof
508            Zürich, Bäckeranlage
564    Zürich, Militär-/Langstrasse
...                             ...
18067          Zürich, Freiestrasse
18099             Zürich, Klusplatz
18161        Zürich, Hölderlinsteig
18269              Zürich, Klosbach
18320            Zürich, Hofstrasse

[193 rows x 1 columns]

In [41]:
tr[tr['itcs_stopName']!='-']['itcs_stopName'].unique()

array(['Zürich, Herdernstrasse', 'Zürich, Hardplatz',
       'Zürich, Güterbahnhof', 'Zürich, Bäckeranlage',
       'Zürich, Militär-/Langstrasse', 'Zürich, Kanonengasse',
       'Zürich, Sihlpost / HB', 'Zürich, Löwenplatz',
       'Zürich, Bahnhofplatz/HB', 'Zürich, Central', 'Zürich, Neumarkt',
       'Zürich, Kunsthaus', 'Zürich, Sprecherstrasse',
       'Zürich, Kreuzplatz', 'Zürich, Signaustrasse',
       'Zürich, Hegibachplatz', 'Zürich, Botanischer Garten',
       'Zürich, Höschgasse', 'Zürich, Fröhlichstrasse',
       'Zürich, Wildbachstrasse', 'Zürich, Bahnhof Tiefenbrunnen',
       'Zürich, Freiestrasse', 'Zürich, Klusplatz',
       'Zürich, Hölderlinsteig', 'Zürich, Klosbach', 'Zürich, Hofstrasse',
       'Zürich, Kirche Fluntern', 'Zürich, Hinterbergstrasse',
       'Zürich, Spyriplatz', 'Zürich, Bethanien', 'Zürich, Toblerplatz'],
      dtype=object)

In [42]:
len(tr[tr['itcs_stopName']!='-']['itcs_stopName'].unique())

31

In [43]:
#trips['StopList']=0
#trips['StopList_weights']=0

In [44]:
#%timeit
#nome=trips['name']
#for i in range(len(trips)):
#    nome[i]='ZTBUS_compressed/'+nome[i]+'.csv'
#    a=pd.read_csv(nome[i], usecols=['itcs_stopName'])
#    a=a.drop(a[a['itcs_stopName']=='-'].index)
#    stoplist=a['itcs_stopName'].unique()
    # stoplist=a[a['itcs_stopName']!='-']['itcs_stopName'].unique()
#    trips['StopList'][i]=stoplist
#    weights=[]
#    for stop in stoplist:
#        weights.append(str(a[a['itcs_stopName']==stop].count().item()))
#    trips['StopList_weights'][i] = weights

In [45]:
#trips

In [46]:
#trips['StopList'][0]==test[test['itcs_stopName']!='-']['itcs_stopName'].unique()

In [47]:
#trips['StopList'][0]==trips['StopList'][1][0:31]

In [48]:
#trips['StopList'][0].size

In [49]:
#trips['StopList']

In [50]:
#int(trips['StopList_weights'][0][0])

In [51]:
#L=0
#for i in range(len(trips)):
#    if L<trips['StopList'][i].size:
#        L=trips['StopList'][i].size
#L

In [52]:
#def check_value (value, listL):
#    for i in range(len(listL)):
#        if value==listL[i]:
#            return i
#    return -1

In [53]:
#value=0
#listL=[1,2,3,4,5,6,0,1,2,4,6,8,6]
#print(check_value (value, listL))
#listL=[1,2,3,4,5,6,1,2,4,6,8,6]
#print(check_value (value, listL))
#listL=['a',2,'ciao',4,5,6,1,2,4,6,8,6]
#print(check_value (value, listL))
#listL=['a',2,'ciao',4,5,6,1,2,4,0,8,6]
#print(check_value (value, listL))
#value='ciao'
#print(check_value (value, listL))

In [54]:
#total=0
#for i in range(len(trips)):
    #if len(trips['StopList'][i])!=len(trips['StopList_weights']) : 
#        print(len(trips['StopList'][i]), len(trips['StopList_weights'][i]), i)

In [55]:
#couples=[]
#stationlist=[]
#for j in range(trips['StopList'][0].size):
#    value=trips['StopList'][0][j]
#    k=int(trips['StopList_weights'][0][j])
#    a=trips['itcs_numberOfPassengers_mean'][0]*k
#    for i in range(len(trips)):
#        listL=trips['StopList'][i]
#        if check_value (value, listL)!=-1:
#            k+=int(trips['StopList_weights'][check_value (value, listL)][j])
#            a+=trips['itcs_numberOfPassengers_mean'][i]*int(trips['StopList_weights'][check_value (value, listL)][j])
#    couple=(value, a/k)
#    if a/k!=0:
#        couples.append(couple)
#    stationlist.append(value)

In [56]:
#couples

In [57]:
nome=trips['name']
for i in range(len(trips)):
    nome[i]='ZTBUS/'+nome[i]+'.csv'

C:\Users\maria\AppData\Local\Temp\ipykernel_1152\783282351.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nome[i]='ZTBUS/'+nome[i]+'.csv'
C:\Users\maria\AppData\Local\Temp\ipykernel_1152\783282351.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nome[i]='ZTBUS/'+nome[i]+'.csv'
C:\Users\maria\AppData\Local\Temp\ipykernel_1152\783282351.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nome[i]='ZTBUS/'+nome[i]+'.csv'
C:\Users\m

C:\Users\maria\AppData\Local\Temp\ipykernel_1152\783282351.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nome[i]='ZTBUS/'+nome[i]+'.csv'
C:\Users\maria\AppData\Local\Temp\ipykernel_1152\783282351.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nome[i]='ZTBUS/'+nome[i]+'.csv'
C:\Users\maria\AppData\Local\Temp\ipykernel_1152\783282351.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nome[i]='ZTBUS/'+nome[i]+'.csv'
C:\Users\m

C:\Users\maria\AppData\Local\Temp\ipykernel_1152\783282351.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nome[i]='ZTBUS/'+nome[i]+'.csv'
C:\Users\maria\AppData\Local\Temp\ipykernel_1152\783282351.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nome[i]='ZTBUS/'+nome[i]+'.csv'
C:\Users\maria\AppData\Local\Temp\ipykernel_1152\783282351.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nome[i]='ZTBUS/'+nome[i]+'.csv'
C:\Users\m

In [58]:
nome

0       ZTBUS/B183_2019-04-30_03-18-56_2019-04-30_08-4...
1       ZTBUS/B183_2019-04-30_13-22-07_2019-04-30_17-5...
2       ZTBUS/B183_2019-05-01_05-58-51_2019-05-01_22-3...
3       ZTBUS/B183_2019-05-03_02-50-21_2019-05-03_05-5...
4       ZTBUS/B183_2019-05-03_15-41-57_2019-05-03_23-0...
                              ...                        
1404    ZTBUS/B208_2022-12-06_14-43-49_2022-12-06_18-2...
1405    ZTBUS/B208_2022-12-07_05-13-02_2022-12-07_19-1...
1406    ZTBUS/B208_2022-12-08_05-22-20_2022-12-08_18-3...
1407    ZTBUS/B208_2022-12-09_23-55-12_2022-12-10_03-2...
1408    ZTBUS/B208_2022-12-10_23-34-46_2022-12-11_03-2...
Name: name, Length: 1409, dtype: object

In [59]:
previous=pd.read_csv(nome[0])
previous=previous.drop(previous[previous['itcs_stopName']=='-'].index)
for i in range(len(trips)):
    if i%83==0:
        print("siamo a ", i , "ancora poco: solo ", len(trips)-i, "righe")
    if i+1==len(trips):
        break
    a=pd.read_csv(nome[i+1])
    a=a.drop(a[a['itcs_stopName']=='-'].index)
    frames=[previous, a]
    previous=pd.concat(frames)

siamo a  0 ancora poco: solo  1409 righe
siamo a  83 ancora poco: solo  1326 righe
siamo a  166 ancora poco: solo  1243 righe
siamo a  249 ancora poco: solo  1160 righe
siamo a  332 ancora poco: solo  1077 righe
siamo a  415 ancora poco: solo  994 righe
siamo a  498 ancora poco: solo  911 righe
siamo a  581 ancora poco: solo  828 righe
siamo a  664 ancora poco: solo  745 righe
siamo a  747 ancora poco: solo  662 righe
siamo a  830 ancora poco: solo  579 righe
siamo a  913 ancora poco: solo  496 righe
siamo a  996 ancora poco: solo  413 righe
siamo a  1079 ancora poco: solo  330 righe
siamo a  1162 ancora poco: solo  247 righe
siamo a  1245 ancora poco: solo  164 righe
siamo a  1328 ancora poco: solo  81 righe


In [60]:
previous

time_iso   time_unix  electric_powerDemand  gnss_altitude  \
246    2019-04-30T03:23:02Z  1556594582             46853.790       423.9126   
322    2019-04-30T03:24:18Z  1556594658             54698.170            NaN   
406    2019-04-30T03:25:42Z  1556594742             33962.440            NaN   
508    2019-04-30T03:27:24Z  1556594844             59833.090            NaN   
564    2019-04-30T03:28:20Z  1556594900             11191.580            NaN   
...                     ...         ...                   ...            ...   
12542  2022-12-11T03:03:48Z  1670727828             12677.750       464.3928   
12591  2022-12-11T03:04:37Z  1670727877              5376.588       458.3075   
12636  2022-12-11T03:05:22Z  1670727922            -55071.570       456.9075   
12674  2022-12-11T03:06:00Z  1670727960            -22885.870       452.2150   
12764  2022-12-11T03:07:30Z  1670728050             18520.080       445.0924   

       gnss_course  gnss_latitude  gnss_longitude itcs_busRoute  \
246       1.303761       0.827037        0.148511             -   
322            NaN            NaN             NaN             -   
406            NaN            NaN             NaN             -   
508            NaN            NaN             NaN             -   
564            NaN            NaN             NaN             -   
...            ...            ...             ...           ...   
12542     0.000000       0.827626        0.148469            N2   
12591     0.000000       0.827651        0.148482            N2   
12636     2.076901       0.827671        0.148520            N2   
12674     0.357920       0.827700        0.148547            N2   
12764     1.503765       0.827703        0.148697            N2   

       itcs_numberOfPassengers                 itcs_stopName  ...  \
246                        1.0        Zürich, Herdernstrasse  ...   
322                        2.0             Zürich, Hardplatz  ...   
406                        3.0          Zürich, Güterbahnhof  ...   
508                        5.0          Zürich, Bäckeranlage  ...   
564                        6.0  Zürich, Militär-/Langstrasse  ...   
...                        ...                           ...  ...   
12542                      9.0      Zürich, Zehntenhausplatz  ...   
12591                      0.0     Zürich, Bahnhof Affoltern  ...   
12636                      0.0              Zürich, Fronwald  ...   
12674                      0.0        Zürich, Unteraffoltern  ...   
12764                      0.0         Zürich, Schwandenholz  ...   

       odometry_wheelSpeed_mr  odometry_wheelSpeed_rl  odometry_wheelSpeed_rr  \
246                  0.917510                0.917510                0.917510   
322                  1.242539                1.242539                1.242539   
406                  0.610547                0.610547                0.610547   
508                  1.052014                1.052014                1.052014   
564                  0.000000                0.000000                0.000000   
...                       ...                     ...                     ...   
12542                0.000000                0.000000                0.000000   
12591                0.000000                0.000000                0.000000   
12636                8.008564                8.035274                8.016577   
12674                8.735694                8.754341                8.727656   
12764                8.276562                8.267063                8.270338   

       status_doorIsOpen  status_gridIsAvailable  status_haltBrakeIsActive  \
246                  0.0                     1.0                       0.0   
322                  0.0                     1.0                       0.0   
406                  0.0                     1.0                       0.0   
508                  0.0                     1.0                       0.0   
564                  0.0                     1.0             

In [61]:
previous.groupby('itcs_stopName').mean('itcs_numberOfPassengers')['itcs_numberOfPassengers']

itcs_stopName
Zürich, Albisrank                 16.196598
Zürich, Albisriederplatz          25.453398
Zürich, Altes Krematorium         21.980348
Zürich, Bahnhof Affoltern          2.850746
Zürich, Bahnhof Altstetten         7.816580
                                    ...    
Zürich, Zentrum Witikon           17.255102
Zürich, Zweiackerstrasse           7.502203
Zürich, Zwielplatz                 2.049383
Zürich, Zwinglihaus               23.196668
Zürich,Kalkbreite/Bhf.Wiedikon    23.160748
Name: itcs_numberOfPassengers, Length: 149, dtype: float64

In [62]:
len(previous['itcs_stopName'].unique())

149

## 6. For each station, determine the buses that have stopped there at least once.

In [63]:
previous.columns

Index(['time_iso', 'time_unix', 'electric_powerDemand', 'gnss_altitude',
       'gnss_course', 'gnss_latitude', 'gnss_longitude', 'itcs_busRoute',
       'itcs_numberOfPassengers', 'itcs_stopName',
       'odometry_articulationAngle', 'odometry_steeringAngle',
       'odometry_vehicleSpeed', 'odometry_wheelSpeed_fl',
       'odometry_wheelSpeed_fr', 'odometry_wheelSpeed_ml',
       'odometry_wheelSpeed_mr', 'odometry_wheelSpeed_rl',
       'odometry_wheelSpeed_rr', 'status_doorIsOpen', 'status_gridIsAvailable',
       'status_haltBrakeIsActive', 'status_parkBrakeIsActive',
       'temperature_ambient', 'traction_brakePressure',
       'traction_tractionForce'],
      dtype='object')

In [64]:
previous

time_iso   time_unix  electric_powerDemand  gnss_altitude  \
246    2019-04-30T03:23:02Z  1556594582             46853.790       423.9126   
322    2019-04-30T03:24:18Z  1556594658             54698.170            NaN   
406    2019-04-30T03:25:42Z  1556594742             33962.440            NaN   
508    2019-04-30T03:27:24Z  1556594844             59833.090            NaN   
564    2019-04-30T03:28:20Z  1556594900             11191.580            NaN   
...                     ...         ...                   ...            ...   
12542  2022-12-11T03:03:48Z  1670727828             12677.750       464.3928   
12591  2022-12-11T03:04:37Z  1670727877              5376.588       458.3075   
12636  2022-12-11T03:05:22Z  1670727922            -55071.570       456.9075   
12674  2022-12-11T03:06:00Z  1670727960            -22885.870       452.2150   
12764  2022-12-11T03:07:30Z  1670728050             18520.080       445.0924   

       gnss_course  gnss_latitude  gnss_longitude itcs_busRoute  \
246       1.303761       0.827037        0.148511             -   
322            NaN            NaN             NaN             -   
406            NaN            NaN             NaN             -   
508            NaN            NaN             NaN             -   
564            NaN            NaN             NaN             -   
...            ...            ...             ...           ...   
12542     0.000000       0.827626        0.148469            N2   
12591     0.000000       0.827651        0.148482            N2   
12636     2.076901       0.827671        0.148520            N2   
12674     0.357920       0.827700        0.148547            N2   
12764     1.503765       0.827703        0.148697            N2   

       itcs_numberOfPassengers                 itcs_stopName  ...  \
246                        1.0        Zürich, Herdernstrasse  ...   
322                        2.0             Zürich, Hardplatz  ...   
406                        3.0          Zürich, Güterbahnhof  ...   
508                        5.0          Zürich, Bäckeranlage  ...   
564                        6.0  Zürich, Militär-/Langstrasse  ...   
...                        ...                           ...  ...   
12542                      9.0      Zürich, Zehntenhausplatz  ...   
12591                      0.0     Zürich, Bahnhof Affoltern  ...   
12636                      0.0              Zürich, Fronwald  ...   
12674                      0.0        Zürich, Unteraffoltern  ...   
12764                      0.0         Zürich, Schwandenholz  ...   

       odometry_wheelSpeed_mr  odometry_wheelSpeed_rl  odometry_wheelSpeed_rr  \
246                  0.917510                0.917510                0.917510   
322                  1.242539                1.242539                1.242539   
406                  0.610547                0.610547                0.610547   
508                  1.052014                1.052014                1.052014   
564                  0.000000                0.000000                0.000000   
...                       ...                     ...                     ...   
12542                0.000000                0.000000                0.000000   
12591                0.000000                0.000000                0.000000   
12636                8.008564                8.035274                8.016577   
12674                8.735694                8.754341                8.727656   
12764                8.276562                8.267063                8.270338   

       status_doorIsOpen  status_gridIsAvailable  status_haltBrakeIsActive  \
246                  0.0                     1.0                       0.0   
322                  0.0                     1.0                       0.0   
406                  0.0                     1.0                       0.0   
508                  0.0                     1.0                       0.0   
564                  0.0                     1.0             

In [65]:
trips.head(5)

name  busNumber  \
0  ZTBUS/B183_2019-04-30_03-18-56_2019-04-30_08-4...        183   
1  ZTBUS/B183_2019-04-30_13-22-07_2019-04-30_17-5...        183   
2  ZTBUS/B183_2019-05-01_05-58-51_2019-05-01_22-3...        183   
3  ZTBUS/B183_2019-05-03_02-50-21_2019-05-03_05-5...        183   
4  ZTBUS/B183_2019-05-03_15-41-57_2019-05-03_23-0...        183   

          startTime_iso  startTime_unix           endTime_iso  endTime_unix  \
0  2019-04-30T03:18:56Z      1556594336  2019-04-30T08:44:20Z    1556613860   
1  2019-04-30T13:22:07Z      1556630527  2019-04-30T17:54:02Z    1556646842   
2  2019-05-01T05:58:51Z      1556690331  2019-05-01T22:32:30Z    1556749950   
3  2019-05-03T02:50:21Z      1556851821  2019-05-03T05:53:20Z    1556862800   
4  2019-05-03T15:41:57Z      1556898117  2019-05-03T23:06:24Z    1556924784   

   drivenDistance busRoute  energyConsumption  itcs_numberOfPassengers_mean  \
0        77213.87        -       4.785852e+08                      5.538860   
1        59029.60       31       4.022585e+08                     33.114580   
2       240900.40       33       1.445733e+09                     19.689140   
3        42565.48        -       2.819867e+08                      1.685185   
4       125277.20       72       6.207258e+08                     23.753570   

   itcs_numberOfPassengers_min  itcs_numberOfPassengers_max  \
0                            0                           20   
1                            4                           74   
2                            0                           55   
3                            0                            8   
4                            1                           67   

   status_gridIsAvailable_mean  temperature_ambient_mean  \
0                     0.740640                  282.3780   
1                     0.855234                  287.5443   
2                     0.777860                  288.7490   
3                     0.767122                  282.4129   
4                     0.907342                  284.7325   

   temperature_ambient_min  temperature_ambient_max   EnergyRatio  
0                   281.15                   293.15  8.640500e+07  
1                   285.15                   293.15  1.214747e+07  
2                   280.15                   294.15  7.342794e+07  
3                   281.15                   292.15  1.673328e+08  
4                   282.15                   287.15  2.613190e+07

in previous 
-> aggiungo colonna con numerobus 
-> raggruppi per stopname e in caso per numerobus 
-> conti se in ogni gruppetto c'è solo 183 oopure 208 oppure entrambi 

In [66]:
previous.head()

time_iso   time_unix  electric_powerDemand  gnss_altitude  \
246  2019-04-30T03:23:02Z  1556594582              46853.79       423.9126   
322  2019-04-30T03:24:18Z  1556594658              54698.17            NaN   
406  2019-04-30T03:25:42Z  1556594742              33962.44            NaN   
508  2019-04-30T03:27:24Z  1556594844              59833.09            NaN   
564  2019-04-30T03:28:20Z  1556594900              11191.58            NaN   

     gnss_course  gnss_latitude  gnss_longitude itcs_busRoute  \
246     1.303761       0.827037        0.148511             -   
322          NaN            NaN             NaN             -   
406          NaN            NaN             NaN             -   
508          NaN            NaN             NaN             -   
564          NaN            NaN             NaN             -   

     itcs_numberOfPassengers                 itcs_stopName  ...  \
246                      1.0        Zürich, Herdernstrasse  ...   
322                      2.0             Zürich, Hardplatz  ...   
406                      3.0          Zürich, Güterbahnhof  ...   
508                      5.0          Zürich, Bäckeranlage  ...   
564                      6.0  Zürich, Militär-/Langstrasse  ...   

     odometry_wheelSpeed_mr  odometry_wheelSpeed_rl  odometry_wheelSpeed_rr  \
246                0.917510                0.917510                0.917510   
322                1.242539                1.242539                1.242539   
406                0.610547                0.610547                0.610547   
508                1.052014                1.052014                1.052014   
564                0.000000                0.000000                0.000000   

     status_doorIsOpen  status_gridIsAvailable  status_haltBrakeIsActive  \
246                0.0                     1.0                       0.0   
322                0.0                     1.0                       0.0   
406                0.0                     1.0                       0.0   
508                0.0                     1.0                       0.0   
564                0.0                     1.0                       0.0   

     status_parkBrakeIsActive  temperature_ambient  traction_brakePressure  \
246                       0.0               291.15                     0.0   
322                       0.0               288.15                     0.0   
406                       0.0               287.15                     0.0   
508                       0.0               287.15                     0.0   
564                       0.0               286.15                     0.0   

     traction_tractionForce  
246               14819.530  
322               16939.120  
406                6230.346  
508               20058.230  
564               12064.050  

[5 rows x 26 columns]

In [67]:
prova = pd.read_csv('ZTBus/B183_2019-04-30_03-18-56_2019-04-30_08-44-20.csv')
prova.head(3)

time_iso   time_unix  electric_powerDemand  gnss_altitude  \
0  2019-04-30T03:18:56Z  1556594336            -13.845510            NaN   
1  2019-04-30T03:18:57Z  1556594337             -3.849362            NaN   
2  2019-04-30T03:18:58Z  1556594338             -0.672331            NaN   

   gnss_course  gnss_latitude  gnss_longitude itcs_busRoute  \
0          NaN            NaN             NaN             -   
1          NaN            NaN             NaN             -   
2          NaN            NaN             NaN             -   

   itcs_numberOfPassengers itcs_stopName  ...  odometry_wheelSpeed_mr  \
0                      NaN             -  ...                     0.0   
1                      NaN             -  ...                     0.0   
2                      NaN             -  ...                     0.0   

   odometry_wheelSpeed_rl  odometry_wheelSpeed_rr  status_doorIsOpen  \
0                     0.0                     0.0                  1   
1                     0.0                     0.0                  1   
2                     0.0                     0.0                  1   

   status_gridIsAvailable  status_haltBrakeIsActive  status_parkBrakeIsActive  \
0                       1                         0                         0   
1                       1                         0                         0   
2                       1                         0                         0   

   temperature_ambient  traction_brakePressure  traction_tractionForce  
0             293.1500                251666.7                     0.0  
1             292.3688                254876.2                     0.0  
2             292.9310                251783.3                     0.0  

[3 rows x 26 columns]

In [68]:
nomi = trips['name']
nomi

0       ZTBUS/B183_2019-04-30_03-18-56_2019-04-30_08-4...
1       ZTBUS/B183_2019-04-30_13-22-07_2019-04-30_17-5...
2       ZTBUS/B183_2019-05-01_05-58-51_2019-05-01_22-3...
3       ZTBUS/B183_2019-05-03_02-50-21_2019-05-03_05-5...
4       ZTBUS/B183_2019-05-03_15-41-57_2019-05-03_23-0...
                              ...                        
1404    ZTBUS/B208_2022-12-06_14-43-49_2022-12-06_18-2...
1405    ZTBUS/B208_2022-12-07_05-13-02_2022-12-07_19-1...
1406    ZTBUS/B208_2022-12-08_05-22-20_2022-12-08_18-3...
1407    ZTBUS/B208_2022-12-09_23-55-12_2022-12-10_03-2...
1408    ZTBUS/B208_2022-12-10_23-34-46_2022-12-11_03-2...
Name: name, Length: 1409, dtype: object

In [69]:
nome

0       ZTBUS/B183_2019-04-30_03-18-56_2019-04-30_08-4...
1       ZTBUS/B183_2019-04-30_13-22-07_2019-04-30_17-5...
2       ZTBUS/B183_2019-05-01_05-58-51_2019-05-01_22-3...
3       ZTBUS/B183_2019-05-03_02-50-21_2019-05-03_05-5...
4       ZTBUS/B183_2019-05-03_15-41-57_2019-05-03_23-0...
                              ...                        
1404    ZTBUS/B208_2022-12-06_14-43-49_2022-12-06_18-2...
1405    ZTBUS/B208_2022-12-07_05-13-02_2022-12-07_19-1...
1406    ZTBUS/B208_2022-12-08_05-22-20_2022-12-08_18-3...
1407    ZTBUS/B208_2022-12-09_23-55-12_2022-12-10_03-2...
1408    ZTBUS/B208_2022-12-10_23-34-46_2022-12-11_03-2...
Name: name, Length: 1409, dtype: object

In [70]:
prova['numerbus'] = 0

In [71]:
prova.head(2)

time_iso   time_unix  electric_powerDemand  gnss_altitude  \
0  2019-04-30T03:18:56Z  1556594336            -13.845510            NaN   
1  2019-04-30T03:18:57Z  1556594337             -3.849362            NaN   

   gnss_course  gnss_latitude  gnss_longitude itcs_busRoute  \
0          NaN            NaN             NaN             -   
1          NaN            NaN             NaN             -   

   itcs_numberOfPassengers itcs_stopName  ...  odometry_wheelSpeed_rl  \
0                      NaN             -  ...                     0.0   
1                      NaN             -  ...                     0.0   

   odometry_wheelSpeed_rr  status_doorIsOpen  status_gridIsAvailable  \
0                     0.0                  1                       1   
1                     0.0                  1                       1   

   status_haltBrakeIsActive  status_parkBrakeIsActive  temperature_ambient  \
0                         0                         0             293.1500   
1                         0                         0             292.3688   

   traction_brakePressure  traction_tractionForce  numerbus  
0                251666.7                     0.0         0  
1                254876.2                     0.0         0  

[2 rows x 27 columns]

In [72]:
m = re.match('^ZTBUS/B183',nome[0])
if m:
    prova['numerobus'] = 183
else:
    prova['numerbus'] = 208

In [73]:
prova.head(2)

time_iso   time_unix  electric_powerDemand  gnss_altitude  \
0  2019-04-30T03:18:56Z  1556594336            -13.845510            NaN   
1  2019-04-30T03:18:57Z  1556594337             -3.849362            NaN   

   gnss_course  gnss_latitude  gnss_longitude itcs_busRoute  \
0          NaN            NaN             NaN             -   
1          NaN            NaN             NaN             -   

   itcs_numberOfPassengers itcs_stopName  ...  odometry_wheelSpeed_rr  \
0                      NaN             -  ...                     0.0   
1                      NaN             -  ...                     0.0   

   status_doorIsOpen  status_gridIsAvailable  status_haltBrakeIsActive  \
0                  1                       1                         0   
1                  1                       1                         0   

   status_parkBrakeIsActive  temperature_ambient  traction_brakePressure  \
0                         0             293.1500                251666.7   
1                         0             292.3688                254876.2   

   traction_tractionForce  numerbus  numerobus  
0                     0.0         0        183  
1                     0.0         0        183  

[2 rows x 28 columns]

In [74]:
previous['nbus'] = 0
previous.head(2)

time_iso   time_unix  electric_powerDemand  gnss_altitude  \
246  2019-04-30T03:23:02Z  1556594582              46853.79       423.9126   
322  2019-04-30T03:24:18Z  1556594658              54698.17            NaN   

     gnss_course  gnss_latitude  gnss_longitude itcs_busRoute  \
246     1.303761       0.827037        0.148511             -   
322          NaN            NaN             NaN             -   

     itcs_numberOfPassengers           itcs_stopName  ...  \
246                      1.0  Zürich, Herdernstrasse  ...   
322                      2.0       Zürich, Hardplatz  ...   

     odometry_wheelSpeed_rl  odometry_wheelSpeed_rr  status_doorIsOpen  \
246                0.917510                0.917510                0.0   
322                1.242539                1.242539                0.0   

     status_gridIsAvailable  status_haltBrakeIsActive  \
246                     1.0                       0.0   
322                     1.0                       0.0   

     status_parkBrakeIsActive  temperature_ambient  traction_brakePressure  \
246                       0.0               291.15                     0.0   
322                       0.0               288.15                     0.0   

     traction_tractionForce  nbus  
246                14819.53     0  
322                16939.12     0  

[2 rows x 27 columns]

In [75]:
m = re.match('^ZTBUS/B183',nome[0])
if m:
    previous['nbus'] = 183
else:
    previous['nbus'] = 208

In [76]:
previous.head(3)

time_iso   time_unix  electric_powerDemand  gnss_altitude  \
246  2019-04-30T03:23:02Z  1556594582              46853.79       423.9126   
322  2019-04-30T03:24:18Z  1556594658              54698.17            NaN   
406  2019-04-30T03:25:42Z  1556594742              33962.44            NaN   

     gnss_course  gnss_latitude  gnss_longitude itcs_busRoute  \
246     1.303761       0.827037        0.148511             -   
322          NaN            NaN             NaN             -   
406          NaN            NaN             NaN             -   

     itcs_numberOfPassengers           itcs_stopName  ...  \
246                      1.0  Zürich, Herdernstrasse  ...   
322                      2.0       Zürich, Hardplatz  ...   
406                      3.0    Zürich, Güterbahnhof  ...   

     odometry_wheelSpeed_rl  odometry_wheelSpeed_rr  status_doorIsOpen  \
246                0.917510                0.917510                0.0   
322                1.242539                1.242539                0.0   
406                0.610547                0.610547                0.0   

     status_gridIsAvailable  status_haltBrakeIsActive  \
246                     1.0                       0.0   
322                     1.0                       0.0   
406                     1.0                       0.0   

     status_parkBrakeIsActive  temperature_ambient  traction_brakePressure  \
246                       0.0               291.15                     0.0   
322                       0.0               288.15                     0.0   
406                       0.0               287.15                     0.0   

     traction_tractionForce  nbus  
246               14819.530   183  
322               16939.120   183  
406                6230.346   183  

[3 rows x 27 columns]

In [77]:
for i in range(len(trips)):
    m = re.match('ZTBUS/B183',nome[i])
    if m != 'None':
        previous['nbus'] = 183
    else:
        previous['nbus'] = 208

In [78]:
previous

time_iso   time_unix  electric_powerDemand  gnss_altitude  \
246    2019-04-30T03:23:02Z  1556594582             46853.790       423.9126   
322    2019-04-30T03:24:18Z  1556594658             54698.170            NaN   
406    2019-04-30T03:25:42Z  1556594742             33962.440            NaN   
508    2019-04-30T03:27:24Z  1556594844             59833.090            NaN   
564    2019-04-30T03:28:20Z  1556594900             11191.580            NaN   
...                     ...         ...                   ...            ...   
12542  2022-12-11T03:03:48Z  1670727828             12677.750       464.3928   
12591  2022-12-11T03:04:37Z  1670727877              5376.588       458.3075   
12636  2022-12-11T03:05:22Z  1670727922            -55071.570       456.9075   
12674  2022-12-11T03:06:00Z  1670727960            -22885.870       452.2150   
12764  2022-12-11T03:07:30Z  1670728050             18520.080       445.0924   

       gnss_course  gnss_latitude  gnss_longitude itcs_busRoute  \
246       1.303761       0.827037        0.148511             -   
322            NaN            NaN             NaN             -   
406            NaN            NaN             NaN             -   
508            NaN            NaN             NaN             -   
564            NaN            NaN             NaN             -   
...            ...            ...             ...           ...   
12542     0.000000       0.827626        0.148469            N2   
12591     0.000000       0.827651        0.148482            N2   
12636     2.076901       0.827671        0.148520            N2   
12674     0.357920       0.827700        0.148547            N2   
12764     1.503765       0.827703        0.148697            N2   

       itcs_numberOfPassengers                 itcs_stopName  ...  \
246                        1.0        Zürich, Herdernstrasse  ...   
322                        2.0             Zürich, Hardplatz  ...   
406                        3.0          Zürich, Güterbahnhof  ...   
508                        5.0          Zürich, Bäckeranlage  ...   
564                        6.0  Zürich, Militär-/Langstrasse  ...   
...                        ...                           ...  ...   
12542                      9.0      Zürich, Zehntenhausplatz  ...   
12591                      0.0     Zürich, Bahnhof Affoltern  ...   
12636                      0.0              Zürich, Fronwald  ...   
12674                      0.0        Zürich, Unteraffoltern  ...   
12764                      0.0         Zürich, Schwandenholz  ...   

       odometry_wheelSpeed_rl  odometry_wheelSpeed_rr  status_doorIsOpen  \
246                  0.917510                0.917510                0.0   
322                  1.242539                1.242539                0.0   
406                  0.610547                0.610547                0.0   
508                  1.052014                1.052014                0.0   
564                  0.000000                0.000000                0.0   
...                       ...                     ...                ...   
12542                0.000000                0.000000                0.0   
12591                0.000000                0.000000                0.0   
12636                8.035274                8.016577                0.0   
12674                8.754341                8.727656                0.0   
12764                8.267063                8.270338                0.0   

       status_gridIsAvailable  status_haltBrakeIsActive  \
246                       1.0                       0.0   
322                       1.0                       0.0   
406                       1.0                       0.0   
508                       1.0                       0.0   
564                       1.0                       0.0   
...                       ...                       ...   
12542                     0.0                       1.0   
12591                     0.0          

## 9. For each (route, bus) pair, compute the ratio between the overall energy consumption and the overall driven distance.

In [79]:
aggr = trips.groupby(['busRoute','busNumber']).agg({'energyConsumption':'sum', 'drivenDistance':'sum'},ignore_index=True)
aggr

energyConsumption  drivenDistance
busRoute busNumber                                   
-        183             1.245498e+10      1999119.75
31       183             1.387782e+10      2279952.00
         208             5.535865e+09      1046306.68
32       183             1.393320e+10      2256540.10
         208             1.283340e+10      2337064.21
33       183             1.761996e+11     29510481.41
         208             3.474842e+10      6161892.39
46       183             1.026237e+11     18224657.94
         208             2.337366e+10      4186493.70
72       183             1.394368e+11     23637238.45
         208             5.100382e+10      9427138.15
83       183             2.995540e+11     51249813.70
         208             2.907837e+11     49965928.79
N1       183             4.217283e+09       705122.71
         208             2.316328e+09       410691.93
N2       183             7.929098e+09      1390618.35
         208             7.913191e+09      1464008.67
N4       183             4.720068e+09       766868.89
         208             3.152512e+09       519599.97

In [80]:
gr = trips.groupby(['busRoute','busNumber']).sum()[['energyConsumption', 'drivenDistance']]

In [81]:
gr

energyConsumption  drivenDistance
busRoute busNumber                                   
-        183             1.245498e+10      1999119.75
31       183             1.387782e+10      2279952.00
         208             5.535865e+09      1046306.68
32       183             1.393320e+10      2256540.10
         208             1.283340e+10      2337064.21
33       183             1.761996e+11     29510481.41
         208             3.474842e+10      6161892.39
46       183             1.026237e+11     18224657.94
         208             2.337366e+10      4186493.70
72       183             1.394368e+11     23637238.45
         208             5.100382e+10      9427138.15
83       183             2.995540e+11     51249813.70
         208             2.907837e+11     49965928.79
N1       183             4.217283e+09       705122.71
         208             2.316328e+09       410691.93
N2       183             7.929098e+09      1390618.35
         208             7.913191e+09      1464008.67
N4       183             4.720068e+09       766868.89
         208             3.152512e+09       519599.97

In [82]:
gr['ratio'] = gr['energyConsumption'] / gr['drivenDistance']

In [83]:
gr

energyConsumption  drivenDistance        ratio
busRoute busNumber                                                
-        183             1.245498e+10      1999119.75  6230.233181
31       183             1.387782e+10      2279952.00  6086.891785
         208             5.535865e+09      1046306.68  5290.862713
32       183             1.393320e+10      2256540.10  6174.585331
         208             1.283340e+10      2337064.21  5491.247671
33       183             1.761996e+11     29510481.41  5970.747507
         208             3.474842e+10      6161892.39  5639.244537
46       183             1.026237e+11     18224657.94  5631.035959
         208             2.337366e+10      4186493.70  5583.110707
72       183             1.394368e+11     23637238.45  5899.029643
         208             5.100382e+10      9427138.15  5410.318878
83       183             2.995540e+11     51249813.70  5844.977154
         208             2.907837e+11     49965928.79  5819.638891
N1       183             4.217283e+09       705122.71  5980.920966
         208             2.316328e+09       410691.93  5640.061883
N2       183             7.929098e+09      1390618.35  5701.850332
         208             7.913191e+09      1464008.67  5405.152963
N4       183             4.720068e+09       766868.89  6154.986937
         208             3.152512e+09       519599.97  6067.190112

In [84]:
risultato = 1.245498e+10 / 1999119.75
print(risultato)

6230.232080894603


## 10. Starting from the results of the previous point, for each route compute the buses with max and min energy ratio, and save the difference between these ratios in a dataframe

In [85]:
#aggr = trips.groupby('busRoute').agg({'energyConsumption':'sum', 'drivenDistance':'sum'},ignore_index=True)
#aggr

energyConsumption  drivenDistance
busRoute                                   
-              1.245498e+10    1.999120e+06
31             1.941369e+10    3.326259e+06
32             2.676660e+10    4.593604e+06
33             2.109481e+11    3.567237e+07
46             1.259974e+11    2.241115e+07
72             1.904406e+11    3.306438e+07
83             5.903377e+11    1.012157e+08
N1             6.533611e+09    1.115815e+06
N2             1.584229e+10    2.854627e+06
N4             7.872580e+09    1.286469e+06

In [94]:
gr['energyConsumption'] / gr['drivenDistance']

busRoute  busNumber
-         183          6230.233181
31        183          6086.891785
          208          5290.862713
32        183          6174.585331
          208          5491.247671
33        183          5970.747507
          208          5639.244537
46        183          5631.035959
          208          5583.110707
72        183          5899.029643
          208          5410.318878
83        183          5844.977154
          208          5819.638891
N1        183          5980.920966
          208          5640.061883
N2        183          5701.850332
          208          5405.152963
N4        183          6154.986937
          208          6067.190112
dtype: float64

In [101]:
df1 = gr.loc[gr['ratio'].groupby('busRoute').idxmax()]
df1

energyConsumption  drivenDistance        ratio
busRoute busNumber                                                
-        183             1.245498e+10      1999119.75  6230.233181
31       183             1.387782e+10      2279952.00  6086.891785
32       183             1.393320e+10      2256540.10  6174.585331
33       183             1.761996e+11     29510481.41  5970.747507
46       183             1.026237e+11     18224657.94  5631.035959
72       183             1.394368e+11     23637238.45  5899.029643
83       183             2.995540e+11     51249813.70  5844.977154
N1       183             4.217283e+09       705122.71  5980.920966
N2       183             7.929098e+09      1390618.35  5701.850332
N4       183             4.720068e+09       766868.89  6154.986937

In [98]:
df2 = gr.loc[gr['ratio'].groupby('busRoute').idxmin()]
df2

energyConsumption  drivenDistance        ratio
busRoute busNumber                                                
-        183             1.245498e+10      1999119.75  6230.233181
31       208             5.535865e+09      1046306.68  5290.862713
32       208             1.283340e+10      2337064.21  5491.247671
33       208             3.474842e+10      6161892.39  5639.244537
46       208             2.337366e+10      4186493.70  5583.110707
72       208             5.100382e+10      9427138.15  5410.318878
83       208             2.907837e+11     49965928.79  5819.638891
N1       208             2.316328e+09       410691.93  5640.061883
N2       208             7.913191e+09      1464008.67  5405.152963
N4       208             3.152512e+09       519599.97  6067.190112

In [106]:
df_tot = pd.merge(df1, df2, on='busRoute')
df_tot

energyConsumption_x  drivenDistance_x      ratio_x  \
busRoute                                                       
-                1.245498e+10        1999119.75  6230.233181   
31               1.387782e+10        2279952.00  6086.891785   
32               1.393320e+10        2256540.10  6174.585331   
33               1.761996e+11       29510481.41  5970.747507   
46               1.026237e+11       18224657.94  5631.035959   
72               1.394368e+11       23637238.45  5899.029643   
83               2.995540e+11       51249813.70  5844.977154   
N1               4.217283e+09         705122.71  5980.920966   
N2               7.929098e+09        1390618.35  5701.850332   
N4               4.720068e+09         766868.89  6154.986937   

          energyConsumption_y  drivenDistance_y      ratio_y  
busRoute                                                      
-                1.245498e+10        1999119.75  6230.233181  
31               5.535865e+09        1046306.68  5290.862713  
32               1.283340e+10        2337064.21  5491.247671  
33               3.474842e+10        6161892.39  5639.244537  
46               2.337366e+10        4186493.70  5583.110707  
72               5.100382e+10        9427138.15  5410.318878  
83               2.907837e+11       49965928.79  5819.638891  
N1               2.316328e+09         410691.93  5640.061883  
N2               7.913191e+09        1464008.67  5405.152963  
N4               3.152512e+09         519599.97  6067.190112

In [107]:
df_tot['ratio_x'] - df_tot['ratio_y']

busRoute
-       0.000000
31    796.029072
32    683.337660
33    331.502969
46     47.925252
72    488.710765
83     25.338263
N1    340.859083
N2    296.697369
N4     87.796825
dtype: float64

In [102]:
totale = pd.DataFrame(columns=['max_183','min_208','differ'])

In [103]:
totale

Empty DataFrame
Columns: [max_183, min_208, differ]
Index: []

## 14. For each month of the year, compute the average ambient temperature

In [90]:
trips['startTime_iso']=pd.to_datetime(trips['startTime_iso'])

In [91]:
trips['mese']=trips['startTime_iso'].dt.month

In [92]:
trips.head(50)

name  busNumber  \
0   ZTBUS/B183_2019-04-30_03-18-56_2019-04-30_08-4...        183   
1   ZTBUS/B183_2019-04-30_13-22-07_2019-04-30_17-5...        183   
2   ZTBUS/B183_2019-05-01_05-58-51_2019-05-01_22-3...        183   
3   ZTBUS/B183_2019-05-03_02-50-21_2019-05-03_05-5...        183   
4   ZTBUS/B183_2019-05-03_15-41-57_2019-05-03_23-0...        183   
5   ZTBUS/B183_2019-05-05_07-41-02_2019-05-05_23-2...        183   
6   ZTBUS/B183_2019-05-06_03-10-43_2019-05-06_19-2...        183   
7   ZTBUS/B183_2019-05-08_02-54-58_2019-05-08_23-0...        183   
8   ZTBUS/B183_2019-05-09_03-20-44_2019-05-09_22-5...        183   
9   ZTBUS/B183_2019-05-10_03-16-11_2019-05-10_18-5...        183   
10  ZTBUS/B183_2019-05-13_03-10-23_2019-05-13_23-1...        183   
11  ZTBUS/B183_2019-05-14_03-33-11_2019-05-14_12-4...        183   
12  ZTBUS/B183_2019-05-14_13-58-18_2019-05-14_21-0...        183   
13  ZTBUS/B183_2019-05-15_03-04-56_2019-05-15_06-5...        183   
14  ZTBUS/B183_2019-05-15_13-31-00_2019-05-15_22-2...        183   
15  ZTBUS/B183_2019-05-18_07-41-27_2019-05-18_23-0...        183   
16  ZTBUS/B183_2019-05-20_03-07-29_2019-05-20_17-2...        183   
17  ZTBUS/B183_2019-05-21_03-41-23_2019-05-21_22-4...        183   
18  ZTBUS/B183_2019-05-22_02-58-32_2019-05-22_23-0...        183   
19  ZTBUS/B183_2019-05-24_02-52-47_2019-05-24_22-3...        183   
20  ZTBUS/B183_2019-05-25_05-29-18_2019-05-25_22-3...        183   
21  ZTBUS/B183_2019-05-27_03-43-52_2019-05-27_17-3...        183   
22  ZTBUS/B183_2019-05-28_02-46-01_2019-05-28_23-0...        183   
23  ZTBUS/B183_2019-05-29_03-28-27_2019-05-29_21-1...        183   
24  ZTBUS/B183_2019-05-31_04-01-06_2019-05-31_22-5...        183   
25  ZTBUS/B183_2019-06-01_05-05-41_2019-06-01_15-1...        183   
26  ZTBUS/B183_2019-06-01_15-19-38_2019-06-01_23-1...        183   
27  ZTBUS/B183_2019-06-03_02-51-59_2019-06-03_07-1...        183   
28  ZTBUS/B183_2019-06-03_13-30-19_2019-06-03_18-3...        183   
29  ZTBUS/B183_2019-06-04_02-55-46_2019-06-04_10-1...        183   
30  ZTBUS/B183_2019-06-05_02-50-33_2019-06-05_07-1...        183   
31  ZTBUS/B183_2019-06-18_14-01-09_2019-06-18_17-1...        183   
32  ZTBUS/B183_2019-06-19_03-03-04_2019-06-19_21-2...        183   
33  ZTBUS/B183_2019-06-20_03-22-43_2019-06-20_22-5...        183   
34  ZTBUS/B183_2019-06-21_02-54-48_2019-06-21_22-3...        183   
35  ZTBUS/B183_2019-06-22_07-41-10_2019-06-22_23-0...        183   
36  ZTBUS/B183_2019-06-24_03-16-13_2019-06-24_18-5...        183   
37  ZTBUS/B183_2019-06-25_03-22-22_2019-06-25_22-2...        183   
38  ZTBUS/B183_2019-06-26_02-52-08_2019-06-26_12-1...        183   
39  ZTBUS/B183_2019-06-27_13-43-04_2019-06-27_18-1...        183   
40  ZTBUS/B183_2019-06-28_03-22-14_2019-06-28_07-2...        183   
41  ZTBUS/B183_2019-06-29_03-03-47_2019-06-29_17-3...        183   
42  ZTBUS/B183_2019-07-03_02-38-00_2019-07-03_21-2...        183   
43  ZTBUS/B183_2019-07-04_04-09-12_2019-07-04_18-5...        183   
44  ZTBUS/B183_2019-07-07_01-27-19_2019-07-07_23-0...        183   
45  ZTBUS/B183_2019-07-08_03-29-50_2019-07-08_22-4...        183   
46  ZTBUS/B183_2019-07-11_14-01-11_2019-07-11_17-0...        183   
47  ZTBUS/B183_2019-07-12_03-42-45_2019-07-12_23-0...        183   
48  ZTBUS/B183_2019-07-13_07-23-35_2019-07-13_22-3...        183   
49  ZTBUS/B183_2019-07-14_03-13-29_2019-07-14_22-5...        183   

               startTime_iso  startTime_unix           endTime_iso  \
0  2019-04-30 03:18:56+00:00      1556594336  2019-04-30T08:44:20Z   
1  2019-04-30 13:22:07+00:00      1556630527  2019-04-30T17:54:02Z   
2  2019-05-01 05:58:51+00:00      1556690331  2019-05-01T22:32:30Z   
3  2019-05-03 02:50:21+00:00      1556851821  2019-05-03T05:53:20Z   
4  2019-05-03 15:41:57+00:00      1556898117  2019-05-03T23:06:24Z   
5  2019-05-05 07:41:02+00:00      1557042062  2019-05-05T23:20:07Z   
6  2019-05-06 03:10:43+00:00      1557112243  2019-05-06T19:20:34Z   
7  2019-05-08 02:54:

In [93]:
trips.groupby('mese')['temperature_ambient_mean'].mean().reset_index(name='avg_temp')

mese    avg_temp
0      1  278.958474
1      2  280.529928
2      3  283.319157
3      4  286.608216
4      5  290.051582
5      6  296.143297
6      7  297.376967
7      8  295.847943
8      9  292.362038
9     10  287.714072
10    11  281.884420
11    12  279.381561